libkiwix does not compile anymore against libzim7

```
FAILED: test/9f86d08@@searcher@exe/searcher.cpp.o 
ccache c++ -Itest/9f86d08@@searcher@exe -I/usr/local/include -I/usr/include/x86_64-linux-gnu -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -g -DGTEST_HAS_PTHREAD=1 -pthread -MD -MQ 'test/9f86d08@@searcher@exe/searcher.cpp.o' -MF 'test/9f86d08@@searcher@exe/searcher.cpp.o.d' -o 'test/9f86d08@@searcher@exe/searcher.cpp.o' -c ../test/searcher.cpp
../test/searcher.cpp: In member function ‘virtual void kiwix::Searcher_incrementalRange_Test::TestBody()’:
../test/searcher.cpp:30:30: error: no matching function for call to ‘zim::Query::setQuery(const char [4], bool)’
   30 |   query.setQuery("ray", false);
      |                              ^
In file included from ../test/../include/searcher.h:33,
                 from ../test/searcher.cpp:2:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
[36/61] Compiling C++ object 'src/25a6634@@kiwix@sha/reader.cpp.o'.
FAILED: src/25a6634@@kiwix@sha/reader.cpp.o 
ccache c++ -Isrc/25a6634@@kiwix@sha -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/local/include -I/usr/include/x86_64-linux-gnu -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -g -fPIC -pthread -MD -MQ 'src/25a6634@@kiwix@sha/reader.cpp.o' -MF 'src/25a6634@@kiwix@sha/reader.cpp.o.d' -o 'src/25a6634@@kiwix@sha/reader.cpp.o' -c ../src/reader.cpp
../src/reader.cpp: In member function ‘bool kiwix::Reader::searchSuggestionsSmart(const string&, unsigned int, kiwix::SuggestionsList_t&)’:
../src/reader.cpp:386:40: error: no matching function for call to ‘zim::Query::setQuery(const string&, bool)’
  386 |   suggestionQuery.setQuery(prefix, true);
      |                                        ^
In file included from ../src/reader.cpp:23:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
[39/61] Compiling C++ object 'src/25a6634@@kiwix@sha/searcher.cpp.o'.
FAILED: src/25a6634@@kiwix@sha/searcher.cpp.o 
ccache c++ -Isrc/25a6634@@kiwix@sha -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/local/include -I/usr/include/x86_64-linux-gnu -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -g -fPIC -pthread -MD -MQ 'src/25a6634@@kiwix@sha/searcher.cpp.o' -MF 'src/25a6634@@kiwix@sha/searcher.cpp.o.d' -o 'src/25a6634@@kiwix@sha/searcher.cpp.o' -c ../src/searcher.cpp
../src/searcher.cpp: In member function ‘void kiwix::Searcher::search(const string&, unsigned int, unsigned int, bool)’:
../src/searcher.cpp:123:43: error: no matching function for call to ‘zim::Query::setQuery(std::string&, bool)’
  123 |     query.setQuery(unaccentedSearch, false);
      |                                           ^
In file included from ../include/searcher.h:33,
                 from ../src/searcher.cpp:22:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
../src/searcher.cpp:124:11: error: ‘class zim::Query’ has no member named ‘setVerbose’
  124 |     query.setVerbose(verbose);
      |           ^~~~~~~~~~
../src/searcher.cpp: In member function ‘void kiwix::Searcher::geo_search(float, float, float, unsigned int, unsigned int, bool)’:
../src/searcher.cpp:164:9: error: ‘class zim::Query’ has no member named ‘setVerbose’
  164 |   query.setVerbose(verbose);
      |         ^~~~~~~~~~
../src/searcher.cpp:165:27: error: no matching function for call to ‘zim::Query::setQuery(const char [1], bool)’
  165 |   query.setQuery("", false);
      |                           ^
In file included from ../include/searcher.h:33,
                 from ../src/searcher.cpp:22:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
../src/searcher.cpp: In member function ‘void kiwix::Searcher::suggestions(std::string&, bool)’:
../src/searcher.cpp:219:9: error: ‘class zim::Query’ has no member named ‘setVerbose’
  219 |   query.setVerbose(verbose);
      |         ^~~~~~~~~~
../src/searcher.cpp:220:40: error: no matching function for call to ‘zim::Query::setQuery(std::string&, bool)’
  220 |   query.setQuery(unaccentedSearch, true);
      |                                        ^
In file included from ../include/searcher.h:33,
                 from ../src/searcher.cpp:22:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
[41/61] Compiling C++ object 'src/25a6634@@kiwix@sha/server_internalServer.cpp.o'.
FAILED: src/25a6634@@kiwix@sha/server_internalServer.cpp.o 
ccache c++ -Isrc/25a6634@@kiwix@sha -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/local/include -I/usr/include/x86_64-linux-gnu -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -g -fPIC -pthread -MD -MQ 'src/25a6634@@kiwix@sha/server_internalServer.cpp.o' -MF 'src/25a6634@@kiwix@sha/server_internalServer.cpp.o.d' -o 'src/25a6634@@kiwix@sha/server_internalServer.cpp.o' -c ../src/server/internalServer.cpp
../src/server/internalServer.cpp: In function ‘kiwix::SuggestionsList_t kiwix::getSuggestions(const zim::Archive*, const string&, int)’:
../src/server/internalServer.cpp:353:47: error: no matching function for call to ‘zim::Query::setQuery(const string&, bool)’
  353 |     suggestionQuery.setQuery(queryString, true);
      |                                               ^
In file included from ../include/searcher.h:33,
                 from ../src/server/internalServer.cpp:54:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
../src/server/internalServer.cpp: In member function ‘std::unique_ptr<kiwix::Response> kiwix::InternalServer::handle_search(const kiwix::RequestContext&)’:
../src/server/internalServer.cpp:613:13: error: ‘class zim::Query’ has no member named ‘setVerbose’
  613 |       query.setVerbose(m_verbose.load());
      |             ^~~~~~~~~~
../src/server/internalServer.cpp:614:31: error: no matching function for call to ‘zim::Query::setQuery(const char [1], bool)’
  614 |       query.setQuery("", false);
      |                               ^
In file included from ../include/searcher.h:33,
                 from ../src/server/internalServer.cpp:54:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
../src/server/internalServer.cpp:623:40: error: no matching function for call to ‘zim::Query::setQuery(std::string&, bool)’
  623 |       query.setQuery(queryString, false);
      |                                        ^
In file included from ../include/searcher.h:33,
                 from ../src/server/internalServer.cpp:54:
/usr/local/include/zim/search.h:129:12: note: candidate: ‘zim::Query& zim::Query::setQuery(const string&)’
  129 |     Query& setQuery(const std::string& query);
      |            ^~~~~~~~
/usr/local/include/zim/search.h:129:12: note:   candidate expects 1 argument, 2 provided
../src/server/internalServer.cpp:624:13: error: ‘class zim::Query’ has no member named ‘setVerbose’
  624 |       query.setVerbose(m_verbose.load());
      |             ^~~~~~~~~~
[43/61] Compiling C++ object 'src/25a6634@@kiwix@sha/search_renderer.cpp.o'.
ninja: build stopped: subcommand failed.
```

--- a/include/searcher.h
+++ b/include/searcher.h
@@ -52,6 +52,7 @@ class Result
 };
 
 struct SearcherInternal;
+struct SuggestionInternal;
 /**
  * The Searcher class is reponsible to do different kind of search using the
  * fulltext index.
@@ -160,6 +161,7 @@ class Searcher
 
   std::vector<Reader*> readers;
   std::unique_ptr<SearcherInternal> internal;
+  std::unique_ptr<SuggestionInternal> suggestionInternal;
   std::string searchPattern;
   unsigned int estimatedResultCount;
   unsigned int resultStart;

--- a/src/reader.cpp
+++ b/src/reader.cpp
@@ -21,6 +21,7 @@
 #include <time.h>
 
 #include <zim/search.h>
+#include <zim/suggestion.h>
 #include <zim/item.h>
 #include <zim/error.h>
 
@@ -377,35 +378,36 @@ bool Reader::searchSuggestionsSmart(const string& prefix,
                                     SuggestionsList_t& results)
 {
   std::vector<std::string> variants = this->getTitleVariants(prefix);
-  bool retVal = false;
 
-  /* Try to search in the title using fulltext search database */
+  auto suggestionSearcher = zim::SuggestionSearcher(*zimArchive);
 
-  auto suggestionSearcher = zim::Searcher(*zimArchive);
-  zim::Query suggestionQuery;
-  suggestionQuery.setQuery(prefix, true);
-  auto suggestionSearch = suggestionSearcher.search(suggestionQuery);
-
-  if (suggestionSearch.getEstimatedMatches()) {
+  if (zimArchive->hasTitleIndex()) {
+    auto suggestionSearch = suggestionSearcher.suggest(prefix);
     const auto suggestions = suggestionSearch.getResults(0, suggestionsCount);
-    for (auto current = suggestions.begin();
-         current != suggestions.end();
-         current++) {
+    for (auto current : suggestions) {
       SuggestionItem suggestion(current.getTitle(), kiwix::normalize(current.getTitle()),
                                 current.getPath(), current.getSnippet());
       results.push_back(suggestion);
     }
-    retVal = true;
   } else {
+    // Check some of the variants of the prefix
     for (std::vector<std::string>::iterator variantsItr = variants.begin();
          variantsItr != variants.end();
          variantsItr++) {
-      retVal = this->searchSuggestions(*variantsItr, suggestionsCount, results)
-               || retVal;
+      auto suggestionSearch = suggestionSearcher.suggest(*variantsItr);
+      for (auto current : suggestionSearch.getResults(0, suggestionsCount)) {
+        if (results.size() >= suggestionsCount) {
+          break;
+        }
+
+        SuggestionItem suggestion(current.getTitle(), kiwix::normalize(current.getTitle()),
+                                current.getPath(), current.getSnippet());
+        results.push_back(suggestion);
+      }
     }
   }
 
-  return retVal;
+  return results.size() > 0;
 }
 
 /* Get next suggestion */

--- a/src/searcher.cpp
+++ b/src/searcher.cpp
@@ -23,6 +23,7 @@
 #include "reader.h"
 
 #include <zim/search.h>
+#include <zim/suggestion.h>
 
 #include <mustache.hpp>
 #include <cmath>
@@ -37,6 +38,7 @@ class _Result : public Result
 {
  public:
   _Result(zim::SearchResultSet::iterator iterator);
+  _Result(SuggestionItem suggestionItem);
   virtual ~_Result(){};
 
   virtual std::string get_url();
@@ -50,6 +52,8 @@ class _Result : public Result
 
  private:
   zim::SearchResultSet::iterator iterator;
+  SuggestionItem suggestionItem;
+  bool isSuggestion;
 };
 
 struct SearcherInternal : zim::SearchResultSet {
@@ -60,6 +64,14 @@ struct SearcherInternal : zim::SearchResultSet {
   }
 
   zim::SearchResultSet::iterator current_iterator;
+};
+
+struct SuggestionInternal : zim::SuggestionResultSet {
+  explicit SuggestionInternal(const zim::SuggestionResultSet& srs)
+    : zim::SuggestionResultSet(srs),
+      currentIterator(srs.begin()) {}
+
+  zim::SuggestionResultSet::iterator currentIterator;
 };
 
 /* Constructor */
@@ -119,9 +131,9 @@ void Searcher::search(const std::string& search,
       }
     }
     zim::Searcher searcher(archives);
+    searcher.setVerbose(verbose);
     zim::Query query;
-    query.setQuery(unaccentedSearch, false);
-    query.setVerbose(verbose);
+    query.setQuery(unaccentedSearch);
     zim::Search search = searcher.search(query);
     internal.reset(new SearcherInternal(search.getResults(resultStart, maxResultCount)));
     this->estimatedResultCount = search.getEstimatedMatches();
@@ -160,9 +172,9 @@ void Searcher::geo_search(float latitude, float longitude, float distance,
     archives.push_back(*(*current)->getZimArchive());
   }
   zim::Searcher searcher(archives);
+  searcher.setVerbose(verbose);
   zim::Query query;
-  query.setVerbose(verbose);
-  query.setQuery("", false);
+  query.setQuery("");
   query.setGeorange(latitude, longitude, distance);
   zim::Search search = searcher.search(query);
   internal.reset(new SearcherInternal(search.getResults(resultStart, maxResultCount)));
@@ -179,10 +191,20 @@ void Searcher::restart_search()
 
 Result* Searcher::getNextResult()
 {
-  if (internal.get() &&
-             internal->current_iterator != internal->end()) {
+  if (internal.get() && internal->current_iterator != internal->end()) {
     Result* result = new _Result(internal->current_iterator);
     internal->current_iterator++;
+    return result;
+  } else if (suggestionInternal.get() &&
+             suggestionInternal->currentIterator != suggestionInternal->end()) {
+    SuggestionItem item(
+                        suggestionInternal->currentIterator->getTitle(),
+                        normalize(suggestionInternal->currentIterator->getTitle()),
+                        suggestionInternal->currentIterator->getPath(),
+                        suggestionInternal->currentIterator->getSnippet()
+                      );
+    Result* result = new _Result(item);
+    suggestionInternal->currentIterator++;
     return result;
   }
   return NULL;
@@ -209,17 +231,12 @@ void Searcher::suggestions(std::string& searchPattern, const bool verbose)
   this->maxResultCount = 10;
   string unaccentedSearch = removeAccents(searchPattern);
 
-  std::vector<zim::Archive> archives;
-  for (auto current = this->readers.begin(); current != this->readers.end();
-       current++) {
-    archives.push_back(*(*current)->getZimArchive());
-  }
-  zim::Searcher searcher(archives);
-  zim::Query query;
-  query.setVerbose(verbose);
-  query.setQuery(unaccentedSearch, true);
-  zim::Search search = searcher.search(query);
-  internal.reset(new SearcherInternal(search.getResults(resultStart, maxResultCount)));
+  // Multizim suggestion is not supported as of now! taking only one archive
+  zim::Archive archive = *(*this->readers.begin())->getZimArchive();
+  zim::SuggestionSearcher searcher(archive);
+  searcher.setVerbose(verbose);
+  zim::SuggestionSearch search = searcher.suggest(searchPattern);
+  suggestionInternal.reset(new SuggestionInternal(search.getResults(resultStart, maxResultCount)));
   this->estimatedResultCount = search.getEstimatedMatches();
 }
 
@@ -235,40 +252,69 @@ zim::SearchResultSet Searcher::getSearchResultSet()
 }
 
 _Result::_Result(zim::SearchResultSet::iterator iterator)
-    : iterator(iterator)
-{
-}
+    : iterator(iterator),
+      suggestionItem("", "", ""),
+      isSuggestion(false)
+{}
+
+_Result::_Result(SuggestionItem item)
+    : iterator(),
+      suggestionItem(item.getTitle(), item.getNormalizedTitle(), item.getPath(), item.getSnippet()),
+      isSuggestion(true)
+{}
 
 std::string _Result::get_url()
 {
+  if (isSuggestion) {
+    return suggestionItem.getPath();
+  }
   return iterator.getPath();
 }
 std::string _Result::get_title()
 {
+  if (isSuggestion) {
+    return suggestionItem.getTitle();
+  }
   return iterator.getTitle();
 }
 int _Result::get_score()
 {
+  if (isSuggestion) {
+    return 0;
+  }
   return iterator.getScore();
 }
 std::string _Result::get_snippet()
 {
+  if (isSuggestion) {
+    return suggestionItem.getSnippet();
+  }
   return iterator.getSnippet();
 }
 std::string _Result::get_content()
 {
+  if (isSuggestion) return "";
   return iterator->getItem(true).getData();
 }
 int _Result::get_size()
 {
+  if (isSuggestion) {
+    return 0;
+  }
   return iterator.getSize();
 }
 int _Result::get_wordCount()
 {
+  if (isSuggestion) {
+    return 0;
+  }
   return iterator.getWordCount();
 }
 std::string _Result::get_zimId()
 {
+  if (isSuggestion) {
+    return "";
+  }
   std::ostringstream s;
   s << iterator.getZimId();
   return s.str();

--- a/src/server/internalServer.cpp
+++ b/src/server/internalServer.cpp
@@ -58,6 +58,7 @@ extern "C" {
 #include <zim/uuid.h>
 #include <zim/error.h>
 #include <zim/search.h>
+#include <zim/suggestion.h>
 #include <zim/entry.h>
 #include <zim/item.h>
 
@@ -347,14 +348,12 @@ SuggestionsList_t getSuggestions(const zim::Archive* const archive,
                   const std::string& queryString, int suggestionCount)
 {
   SuggestionsList_t suggestions;
+  auto searcher = zim::SuggestionSearcher(*archive);
   if (archive->hasTitleIndex()) {
-    auto searcher = zim::Searcher(*archive);
-    zim::Query suggestionQuery;
-    suggestionQuery.setQuery(queryString, true);
-    auto suggestionSearch = searcher.search(suggestionQuery);
-    auto suggestionResult = suggestionSearch.getResults(0, suggestionCount);
+    auto search = searcher.suggest(queryString);
+    auto srs = search.getResults(0, suggestionCount);
 
-    for (auto it = suggestionResult.begin(); it != suggestionResult.end(); it++) {
+    for (auto it : srs) {
       SuggestionItem suggestion(it.getTitle(), kiwix::normalize(it.getTitle()),
                                 it.getPath(), it.getSnippet());
       suggestions.push_back(suggestion);
@@ -364,9 +363,11 @@ SuggestionsList_t getSuggestions(const zim::Archive* const archive,
     std::vector<std::string> variants = getTitleVariants(queryString);
     int currCount = 0;
     for (auto it = variants.begin(); it != variants.end() && currCount < suggestionCount; it++) {
-      for (auto& entry: archive->findByTitle(*it)) {
-        SuggestionItem suggestion(entry.getTitle(), kiwix::normalize(entry.getTitle()),
-                                  entry.getPath());
+      auto search = searcher.suggest(queryString);
+      auto srs = search.getResults(0, suggestionCount);
+      for (auto it : srs) {
+        SuggestionItem suggestion(it.getTitle(), kiwix::normalize(it.getTitle()),
+                                  it.getPath());
         suggestions.push_back(suggestion);
         currCount++;
       }
@@ -610,8 +611,7 @@ std::unique_ptr<Response> InternalServer::handle_search(const RequestContext& re
         cout << "Performing geo query `" << distance << "&(" << latitude << ";" << longitude << ")'" << endl;
       }
 
-      query.setVerbose(m_verbose.load());
-      query.setQuery("", false);
+      query.setQuery("");
       query.setGeorange(latitude, longitude, distance);
     } else {
       // Execute Ft search
@@ -620,8 +620,7 @@ std::unique_ptr<Response> InternalServer::handle_search(const RequestContext& re
       }
 
       std::string queryString = removeAccents(patternString);
-      query.setQuery(queryString, false);
-      query.setVerbose(m_verbose.load());
+      query.setQuery(queryString);
     }
 
     zim::Search search = searcher->search(query);

--- a/test/searcher.cpp
+++ b/test/searcher.cpp
@@ -22,12 +22,37 @@ TEST(Searcher, search) {
   ASSERT_EQ(result->get_title(), "Wikibooks");
 }
 
+TEST(Searcher, suggestion) {
+  Reader reader("./test/zimfile.zim");
+
+  Searcher searcher;
+  searcher.add_reader(&reader);
+  ASSERT_EQ(searcher.get_reader(0)->getTitle(), reader.getTitle());
+
+  std::string query = "ray";
+  searcher.suggestions(query, true);
+  searcher.restart_search();
+
+  auto result = searcher.getNextResult();
+  ASSERT_EQ(result->get_title(), "Charles, Ray");
+  ASSERT_EQ(result->get_url(), "A/Charles,_Ray");
+  ASSERT_EQ(result->get_snippet(), "Charles, <b>Ray</b>");
+  ASSERT_EQ(result->get_score(), 0);
+  ASSERT_EQ(result->get_content(), "");
+  ASSERT_EQ(result->get_size(), 0);
+  ASSERT_EQ(result->get_wordCount(), 0);
+  ASSERT_EQ(result->get_zimId(), "");
+
+  result = searcher.getNextResult();
+  ASSERT_EQ(result->get_title(), "Ray (film)");
+}
+
 TEST(Searcher, incrementalRange) {
   // Attempt to get 50 results in steps of 5
   zim::Archive archive("./test/zimfile.zim");
   zim::Searcher ftsearcher(archive);
   zim::Query query;
-  query.setQuery("ray", false);
+  query.setQuery("ray");
   auto search = ftsearcher.search(query);
 
   int suggCount = 0;

